### PR TITLE
Refactor celery, add full simulation status and result

### LINKED ISSUE #110: Improve output converter to digest 0-D and 2-D output

### PR DESCRIPTION
**Pull request by p1003**

### CODE REVIEW

File: yaptide/celery/tasks.py [hunk 1]
```diff
@@ -32,7 +28,8 @@ def run_shieldhit(self, param_dict: dict, raw_input_dict: dict):
         # digest dictionary with project data (extracted from JSON file)
         # and generate SHIELD-HIT12A input files
         conv_parser = get_parser_from_str("shieldhit")
-        run_parser(parser=conv_parser, input_data=raw_input_dict, output_dir=tmp_output_path)
+        run_parser(parser=conv_parser, input_data=raw_input_dict,
```
Comment (grzanka) [inline yaptide/celery/tasks.py:31]: why this change ? we should support lines of 120 characters

Comment (p1003) [reply yaptide/celery/tasks.py:31]: It was autocorrect by IDE, rechanged to original

File: yaptide/celery/tasks.py [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+        'message': {
+            'state': task.state
+        }
+    }
+    if task.state == "PENDING":
+        result['message']['status'] = 'Pending...'
+    elif task.state == "PROGRESS":
+        result['message']['status'] = task.state
+        sim_info = read_shieldlog(path_to_file=task.info.get(
+            'path')+'/run_1/shieldhit0001.log')
```
Comment (grzanka) [inline yaptide/celery/tasks.py]:
```suggestion
        sim_info = read_shieldlog(path_to_file=task.info.get('path')+'/run_1/shieldhit0001.log')
```

File: yaptide/celery/tasks.py [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+        if 'result' in task.info:
+            result['message']['result'] = task.info.get('result')
+    else:
+        result['status'] = 'ERROR'
+        result['message']['status'] = str(task.info)
+
+    return result
+
+
+def read_shieldlog(path_to_file: str):
```
Comment (grzanka) [inline yaptide/celery/tasks.py]: Rename to `sim_status_from_logfile` or something similar

File: yaptide/celery/tasks.py [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+    """Function responsible for reading logs of simulation and returning it's current state"""
```
Comment (grzanka) [inline yaptide/celery/tasks.py]:
```suggestion
    """Extracts current simulation state from simulation logfile"""
```

File: yaptide/celery/tasks.py [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+        for line in reader:
+            if not flag:
+                flag = line.lstrip().startswith("Starting transport")
+            else:
+                if line.lstrip().startswith("Primary particle"):
+                    last_result_line = line
+
+        splited = last_result_line.split()
+        sim_info = {
+            'counted': splited[3],
```
Comment (grzanka) [inline yaptide/celery/tasks.py]: Better: `simulated_primaries`

File: yaptide/celery/tasks.py [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+        result['status'] = 'ERROR'
+        result['message']['status'] = str(task.info)
+
+    return result
+
+
+def read_shieldlog(path_to_file: str):
+    """Function responsible for reading logs of simulation and returning it's current state"""
+    with open(path_to_file, 'r') as reader:
+        flag = False
```
Comment (grzanka) [inline yaptide/celery/tasks.py]: Rename to `found_line_which_starts_status_block`

File: yaptide/celery/tasks.py [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+        'status': 'OK',
+        'message': {
+            'state': task.state
+        }
+    }
+    if task.state == "PENDING":
+        result['message']['status'] = 'Pending...'
+    elif task.state == "PROGRESS":
+        result['message']['status'] = task.state
+        sim_info = read_shieldlog(path_to_file=task.info.get(
```
Comment (grzanka) [inline yaptide/celery/tasks.py]: Add a comment mentioning that this is temporary solution, based on reading logs just from a first directory.
Add a TODO comment and issue to make it better

File: yaptide/celery/tasks.py [hunk 8]
```diff
@@ -0,0 +1,181 @@
+from yaptide.celery.worker import celery_app
+
+import sys
+import tempfile
+import os
```
Comment (grzanka) [inline yaptide/celery/tasks.py]:
```suggestion
import os
import sys
import tempfile
```

sort imports

File: yaptide/celery/tasks.py [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+        settings = SimulationSettings(input_path=tmp_dir_path,  # skipcq: PYL-W0612
+                                      simulator_exec_path=None,
+                                      cmdline_opts='')
+
+        runner_obj = SHRunner(jobs=param_dict['jobs'],
+                              keep_workspace_after_run=True,
+                              output_directory=tmp_dir_path)
+
+        self.update_state(state="PROGRESS", meta={"path": tmp_dir_path})
+        isRunOk = runner_obj.run(settings=settings)
```
Comment (grzanka) [inline yaptide/celery/tasks.py]: rename to `is_run_ok`

File: yaptide/celery/tasks.py [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+            # "data" which has unit, name and list of data values
+            page_dict = {
+                "dimensions": page.dimension,
+            }
+            # currently output is returned only when dimension == 1 due to
+            # problems in efficient testing of other dimensions
+            if page.dimension == 0:
+                page_dict["data"] = {
+                    "unit": str(page.unit),
+                    "name": str(page.name),
```
Comment (grzanka) [inline yaptide/celery/tasks.py]: Extract filling unit and name fields to part of the code before `if` block

**Comment by grzanka** [comment]
@p1003 do we need to commit `yaptide/main.db` ?

File: yaptide/celery/tasks.py (hunk 9 above)
Comment (p1003) [reply yaptide/celery/tasks.py]: changed

**Comment by p1003** [comment]
> @p1003 do we need to commit `yaptide/main.db` ?

We don't

File: docker-compose.yml [hunk 11]
```diff
@@ -21,10 +21,10 @@ services:
     build: 
       context: .
       args:
-        SHIELDHIT_PATH: ${SHIELDHIT_PATH}
+        SHIELDHIT_PATH: ${SHIELDHIT_PATH-/shieldhit}
```
Comment (grzanka) [inline docker-compose.yml:24]: What does it mean ? Is it path concatenation ? Or something else ?

File: yaptide/celery/tasks.py [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+
+
+@celery_app.task(bind=True)
+def run_simulation(self, param_dict: dict, raw_input_dict: dict):
+    """Simulation runner"""
+    # create temporary directory
+    with tempfile.TemporaryDirectory() as tmp_dir_path:
+
+        # digest dictionary with project data (extracted from JSON file)
+        # and generate SHIELD-HIT12A input files
```
Comment (grzanka) [inline yaptide/celery/tasks.py]:
```suggestion
        # and generate simulation input files
```

File: yaptide/celery/tasks.py [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+        conv_parser = get_parser_from_str(param_dict['sim_type'])
+        run_parser(parser=conv_parser, input_data=raw_input_dict, output_dir=tmp_dir_path)
+
+        settings = SimulationSettings(input_path=tmp_dir_path,  # skipcq: PYL-W0612
```
Comment (grzanka) [inline yaptide/celery/tasks.py]:
```suggestion
        # we assume here that the simulation executable is available in the PATH env variable and will be automatically discovered by pymchelper
        settings = SimulationSettings(input_path=tmp_dir_path,  # skipcq: PYL-W0612
```

File: yaptide/celery/tasks.py [hunk 14]
```diff
@@ -0,0 +1,181 @@
...
+            raise celery_exceptions.TaskError
+
+        estimators_dict: dict = runner_obj.get_data()
+
+        result: dict = dummy_convert_output(estimators_dict)
+
+        return {'status': 'COMPLETED', 'result': result}
+
+
+def dummy_convert_output(estimators_dict: dict) -> dict:
```
Comment (grzanka) [inline yaptide/celery/tasks.py]: Rename to something non-dummy, like: pymchelper_output_to_json

File: yaptide/celery/tasks.py [hunk 15]
```diff
@@ -0,0 +1,181 @@
...
+    """Dummy function for converting simulation output to dictionary"""
+    if not estimators_dict:
+        return {'message': 'No estimators'}
+
+    # result_dict is the dictionary object, which is later converted to json
+    # to provide readable api response for fronted
+
+    # result_dict contains the list of estimators
```
Comment (grzanka) [inline yaptide/celery/tasks.py]:
```suggestion
    # result_dict is a dictionary, which is later converted to json
    # to provide readable API response for fronted
   # keys in results_dict are estimator names, values are the estimator objects
```

File: yaptide/celery/tasks.py [hunk 16]
```diff
@@ -0,0 +1,181 @@
...
+        # est_dict contains list of pages
+        est_dict = {
+            'name': estimator_key,
+            'pages': [],
+            }
+
+        page: Page   # type is still marked as Never
+        for page in estimator.pages:
+            # currently we are handling for sure only 1-D results
+            # 0-D and 2-D results aren't tested yet, due to testing problems
```
Comment (grzanka) [inline yaptide/celery/tasks.py]: what kind of testing problems do we have ?

File: yaptide/celery/tasks.py [hunk 17]
```diff
@@ -0,0 +1,181 @@
...
+        'status': 'OK',
+        'message': {
+            'state': task.state
+        }
+    }
+    if task.state == "PENDING":
+        result['message']['status'] = 'Pending...'
+    elif task.state == "PROGRESS":
+        result['message']['status'] = 'Calculations in progress...'
+        sim_info = sim_status_from_logfile(path_to_file=task.info.get('path')+'/run_1/shieldhit0001.log')
```
Comment (grzanka) [inline yaptide/celery/tasks.py]:
```suggestion
        sim_info = sim_status_from_logfile(path_to_file=os.path.join(task.info.get('path'), 'run_1', 'shieldhit0001.log')
```

File: yaptide/celery/tasks.py [hunk 18]
```diff
@@ -0,0 +1,181 @@
...
+            if not found_line_which_starts_status_block:
+                # We are searching for lines containing progress info
+                # They are preceded by line starting with "Starting transport"
+                found_line_which_starts_status_block = line.lstrip().startswith("Starting transport")
+            else:
+                # Searching for latest line
+                if line.lstrip().startswith("Primary particle"):
+                    last_result_line = line
+
+        splited = last_result_line.split()
```
Comment (grzanka) [inline yaptide/celery/tasks.py:185]: What if we call this method on a file which was generated in the early phase of simulation. It will not contain any of ""Starting transport" and "Primary particle" lines.. Then `last_result_line` will be empty string and you will not be able to call `splitted[3]`

**Comment by grzanka** [comment]
> > @p1003 do we need to commit `yaptide/main.db` ?
> 
> We don't

But it is still commited in this PR

File: yaptide/celery/tasks.py (hunk 16 above)
Comment (p1003) [reply yaptide/celery/tasks.py]: Testing requires modyfing converter files, which we didn't want to do, that's why they haven't been tested yet, but tomorrow we plan to do it anyway due to little amount time left

File: yaptide/celery/tasks.py (hunk 18 above)
Comment (p1003) [reply yaptide/celery/tasks.py:185]: I'll add some if statement

File: docker-compose.yml (hunk 11 above)
Comment (p1003) [reply docker-compose.yml:24]: @retkiewi

File: yaptide/celery/tasks.py (hunk 14 above)
Comment (p1003) [reply yaptide/celery/tasks.py]: changed

File: yaptide/celery/tasks.py (hunk 7 above)
Comment (p1003) [reply yaptide/celery/tasks.py]: Added in function, because the flow of this code won't change (just function will be)

File: yaptide/celery/tasks.py [hunk 19]
```diff
@@ -0,0 +1,191 @@
...
+        'message': {
+            'state': task.state
+        }
+    }
+    if task.state == "PENDING":
+        result['message']['status'] = 'Pending...'
+    elif task.state == "PROGRESS":
+        result['message']['status'] = 'Calculations in progress...'
+        sim_info = sim_status_from_logfile(path_to_file=os.path.join(
+            task.info.get('path'), 'run_1', 'shieldhit0001.log'))
```
Comment (grzanka) [inline yaptide/celery/tasks.py]:
```suggestion
        sim_info = sim_status_from_logfile(path_to_file=os.path.join(task.info.get('path'), 'run_1', 'shieldhit0001.log'))
```

File: yaptide/celery/tasks.py [hunk 20]
```diff
@@ -0,0 +1,191 @@
...
+    result = {
+        'status': 'OK',
+        'message': {
+            'state': task.state
+        }
+    }
+    if task.state == "PENDING":
+        result['message']['status'] = 'Pending...'
+    elif task.state == "PROGRESS":
+        result['message']['status'] = 'Calculations in progress...'
```
Comment (grzanka) [inline yaptide/celery/tasks.py:125]: I think we have somewhere simulation type being defined.
Can you add here an if statement and call this method only if the simulation type is `shieldhit` ?

File: yaptide/celery/tasks.py [hunk 21]
```diff
@@ -0,0 +1,191 @@
...
+                with open(path, 'r') as reader:
+                    result['message'][path.split('/')[-1]] = reader.read()
+        except FileNotFoundError:
+            result['message']['info'] = "No input present"
+    else:
+        result['message']['info'] = "No input present"
+    return result
+
+
+def sim_status_from_logfile(path_to_file: str):
```
Comment (grzanka) [inline yaptide/celery/tasks.py]: Rename to `sh12a_simulation_status`

File: yaptide/celery/tasks.py [hunk 22]
```diff
@@ -0,0 +1,191 @@
...
+    """Extracts current simulation state from simulation logfile"""
```
Comment (grzanka) [inline yaptide/celery/tasks.py]:
```suggestion
    """Extracts current SHIELD-HIT12A simulation state from first available logfile"""
```

File: yaptide/examples/call_api_example.py [hunk 23]
```diff
@@ -1,39 +1,60 @@
 import requests
 import json
 import time
+import os
 
 
 def call_api():
     """Example backend endpoint call"""
+    # api_post = 'http://localhost:5000/sh/run?sim_type=dummy'
```
Comment (grzanka) [inline yaptide/examples/call_api_example.py]:
```suggestion
```

Remove if not needed

File: yaptide/routes.py [hunk 24]
```diff
@@ -27,7 +26,8 @@ def decorator(f):
         """Determines if the access or refresh token is valid"""
         @wraps(f)
         def wrapper(*args, **kwargs):
-            token: str = request.cookies.get('refresh_token' if isRefresh else 'access_token')
+            token: str = request.cookies.get(
+                'refresh_token' if isRefresh else 'access_token')
```
Comment (grzanka) [inline yaptide/routes.py]:
```suggestion
            token: str = request.cookies.get('refresh_token' if isRefresh else 'access_token')
```

Comment (grzanka) [reply yaptide/routes.py]: also `isRefresh` is no pythonic style. What does it mean ? Shouldn't it rather be `is_refreshed` ?

File: yaptide/routes.py [hunk 25]
```diff
@@ -74,24 +68,31 @@ def post(user: UserModel) -> Union[dict[str, list[str]],
...
-                                   raw_input_dict=json_data)
-
-        return json.dumps({"task_id": task.id}), api_status.HTTP_202_ACCEPTED
+            return make_response({
+                'status': 'ERROR',
+                'message': 'JSON not provided'
+            }, api_status.HTTP_400_BAD_REQUEST)
+
+        task = run_simulation.delay(
+            param_dict=param_dict, raw_input_dict=json_data)
```
Comment (grzanka) [inline yaptide/routes.py]:
```suggestion
        task = run_simulation.delay(param_dict=param_dict, raw_input_dict=json_data)
```

File: yaptide/routes.py [hunk 26]
```diff
@@ -205,26 +215,27 @@ def post():
...
-                return {
+                return make_response({
                     'status': 'ERROR',
                     'message': 'Invalid login or password'
-                }, api_status.HTTP_401_UNAUTHORIZED
+                }, api_status.HTTP_401_UNAUTHORIZED)
 
-            access_token, access_exp = encode_auth_token(user_id=user.id, isRefresh=False)
+            access_token, access_exp = encode_auth_token(
+                user_id=user.id, isRefresh=False)
```
Comment (grzanka) [inline yaptide/routes.py]:
```suggestion
            access_token, access_exp = encode_auth_token(user_id=user.id, isRefresh=False)
```

File: yaptide/routes.py [hunk 27]
```diff
@@ -254,7 +265,8 @@ class UserRefresh(Resource):
     @requires_auth(isRefresh=True)
     def get(user: UserModel):
         """Method refreshing token"""
-        access_token, access_exp = encode_auth_token(user_id=user.id, isRefresh=False)
+        access_token, access_exp = encode_auth_token(
+            user_id=user.id, isRefresh=False)
```
Comment (grzanka) [inline yaptide/routes.py]:
```suggestion
        access_token, access_exp = encode_auth_token(user_id=user.id, isRefresh=False)
```

File: yaptide/celery/tasks.py (hunk 19 above)
Comment (p1003) [reply yaptide/celery/tasks.py]: I'll change it in code myself

File: yaptide/celery/tasks.py (hunk 20 above)
Comment (p1003) [reply yaptide/celery/tasks.py:125]: changed

**Review by grzanka — DISMISSED** [review]
LGTM

**Review by grzanka — APPROVED** [review]
LGTM
